### PR TITLE
When nominating NuGet restore, pass MSBuildProjectExtensionsPath inst…

### LINKED ISSUE #3466: Pass MSBuildProjectExtensionsPath instead of BaseIntermediateOutputPath as part of NuGet restore nomination
Context:

- https://github.com/NuGet/NuGet.Client/pull/2131
- https://github.com/Microsoft/msbuild/issues/1603

Basically, NuGet was writing restore output to `BaseIntermediateOutputPath`, but the .props and .targets were being imported from `MSBuildProjectExtensionsPath`, so if you overrode one but not the other (commonly because you set `BaseIntermediateOutputPath` after the Microsoft common .props were evaluated), the .props written by restore wouldn't be imported.

We've now normalized this in MSBuild and NuGet to use `MSBuildProjectExtensionsPath`.  However, in some cases the path is passed to NuGet from the project system.  So those cases need to be updated to pass `MSBuildProjectExtensionsPath`.

Note that `IVsProjectRestoreInfo` has a `BaseIntermediatePath` property.  In order to avoid breaking the interface, we didn't change the name, but the project system should now set that value to the `MSBuildProjectExtensionsPath` from MSBuild.

[NuGetRestore.xaml](https://github.com/dotnet/project-system/blob/3c760034e97c2551a8ff98fe01e46830d34b4213/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/NuGetRestore.xaml#L45-L47) may also need to be updated, I'm not sure exactly how that works.

@Pilchie I've optimistically set the milestone for this to 15.7.

### PR DESCRIPTION
**Pull request by Pilchie**
…ead of BaseIntermediateOutputPath

Fixes #3466.

<details>
**Customer scenario**

What does the customer do to get into this situation, and why do we think this
is common enough to address in Escrow.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

**Bugs this fixes:** 

(either VSO or GitHub links)

**Workarounds, if any**

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

**Risk**

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

**Performance impact**

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

**Is this a regression from a previous update?**

**Root cause analysis:**

How did we miss it?  What tests are we adding to guard against it in the future?

**How was the bug found?**

(E.g. customer reported it vs. ad hoc testing)
</details>

### CODE REVIEW

**Comment by Pilchie** [comment]
Tagging @dsplaisted to take a look too.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs [hunk 1]
```diff
@@ -87,7 +87,10 @@ internal static class ProjectRestoreInfoBuilder
             return targetFrameworks.Any()
                 ? new ProjectRestoreInfo
                 {
-                    BaseIntermediatePath = baseIntermediatePath,
+                    // NOTE: We pass MSBuildProjectExtensionsPath as BaseIntermediatePath, because people often set
+                    // BaseIntermediatePath too late.  See https://github.com/dotnet/project-system/issues/3466 for
+                    // details.
```
Comment (dsplaisted) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs]: Note that the MSBuild property is `BaseIntermediateOutputPath`, which is not the same as the property name: `BaseIntermediatePath`.  I would reword the comment similar to this:

NOTE: We pass MSBuildProjectExtensionsPath as BaseIntermediatePath instead of using BaseIntermediateOutputPath.  This is because NuGet switched from using BaseIntermediateOutputPath to MSBuildProjectExtensionsPath, since the value of BaseIntermediateOutputPath is often set too late (after *.g.props files would need to have been imported from it).  Instead of modifying the IVsProjectRestoreInfo interface or introducing something like IVsProjectRestoreInfo with an MSBuildProjectExtensionsPath property, we opted to leave the interface the same but change the meaning of its BaseIntermediatePath proprtey.  See https://github.com/dotnet/project-system/issues/3466 for for details.

Comment (Pilchie) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/NuGet/ProjectRestoreInfoBuilder.cs]: Done thanks!

**Comment by Pilchie** [comment]
@dsplaisted - wanted to double check that this will work regardless of the SDK version, right?  We don't need to do any sort of conditioning?

**Comment by davkean** [comment]
Can you expand on that? It's unclear to me if earlier versions of SDKs are broken.

**Comment by dsplaisted** [comment]
Versions 2.1.1xx and prior of the SDK are broken if `MSBuildProjectExtensionsPath` doesn't match `BaseIntermediateOutputPath` (NuGet restore will write the nuget.g.props and .targets files to the latter path, while the MSBuild common targets will try to import them from the prior path).  So there's not any point in trying to make this change to the project system preserve existing behavior for prior SDKs, since the existing behavior is broken anyway.

**Comment by Pilchie** [comment]
Approved to merge via [601526](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/601526)
